### PR TITLE
fix: Leader as transaction ordering service

### PR DESCRIPTION
**Pull request by SamHSmith**
## Description

#4914 

### Linked issue

Closes #{issue_number} 

### Checklist

- [ ] I've read `CONTRIBUTING.md`
- [ ] I've used the standard signed-off commit format (or will squash just before merging)
- [ ] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: core/src/sumeragi/main_loop.rs [hunk 1]
```diff
@@ -383,40 +365,34 @@ impl Sumeragi {
...
-            return None;
-        }
+        BlockCreated {
+            transactions,
+            creation_time_ms,
+        }: BlockCreated,
+    ) -> VotingBlock<'state> {
+        let transactions = transactions
+            .into_iter()
+            .map(|stx| AcceptedTransaction(stx))
```
Comment (mversic) [inline core/src/sumeragi/main_loop.rs:375]:
```suggestion
            .map(|stx| AcceptedTransaction::accept(stx))
```

don't ever skip validation. The inner field of `AcceptedTransaction` should have been private

**Comment by mversic** [comment]
you should follow the style of the code and add a new state to the block state machine. Let's consider this in more detail:
1. remove `categorize_transactions` function
2. make `BlockBuilder::new` -> `BlockBuilder::chain` -> `BlockBuilder::sign` produce `SignedBlock`(not `ValidBlock`)

File: crates/iroha_core/src/sumeragi/message.rs [hunk 2]
```diff
@@ -39,17 +42,12 @@ impl ControlFlowMessage {
 /// `BlockCreated` message structure.
 #[derive(Debug, Clone, Decode, Encode)]
 pub struct BlockCreated {
+    /// The leaders peer id. Helps for cheap pruning of messages from previous rounds.
+    pub leader_signature: SignatureOf<Vec<SignedTransaction>>,
     /// The corresponding block.
-    pub block: SignedBlock,
```
Comment (mversic) [inline crates/iroha_core/src/sumeragi/message.rs:43]: There is no need to change this. Leader can keep sending `SignedBlock` as before

Comment (mversic) [reply crates/iroha_core/src/sumeragi/message.rs:43]: I say this because transaction validation will not change anything except modify the `CommittedTransaction::error` value

Comment (mversic) [reply crates/iroha_core/src/sumeragi/message.rs:43]: IMO leader can send a block where `CommittedTransaction::error` is `None` and update the value later during block validation

**Review by mversic — CHANGES_REQUESTED** [review]

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 3]
```diff
@@ -841,57 +839,74 @@ impl Sumeragi {
...
+
+            let new_voting_block = self
+                .validate_block(
+                    state,
+                    &self.topology,
+                    genesis_account,
+                    block_created_msg,
+                    voting_block,
+                )
+                .expect("We just created this block ourselves, it has to be valid.");
```
Comment (Erigara) [inline crates/iroha_core/src/sumeragi/main_loop.rs:895]: I think this place might panic if any transaction is invalid, isn't it?

Also validators would reject block if any transaction is invalid.

**Review by Erigara — COMMENTED** [review]
Is there is any performance boost measurements?

**Review by mversic — CHANGES_REQUESTED** [review]

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 4]
```diff
@@ -269,7 +267,10 @@ impl Sumeragi {
 
                     // NOTE: By this time genesis block is executed and list of trusted peers is updated
                     self.topology = Topology::new(state_block.world.trusted_peers_ids.clone());
-                    self.commit_block(block, state_block);
+                    self.commit_block(block.clone(), state_block);
+                    self.broadcast_packet(BlockSyncUpdate {
+                        block: block.into(),
+                    });
```
[the remark this answers is not in the document]
Comment (SamHSmith) [reply crates/iroha_core/src/sumeragi/main_loop.rs:273]: I think I mixed up the genesis functions when I was editiing the code. Thanks for spotting this.

File: crates/iroha_core/src/sumeragi/main_loop.rs (hunk 3 above)
Comment (SamHSmith) [reply crates/iroha_core/src/sumeragi/main_loop.rs:895]: My understanding is that the transaction should have been rejected by the queue already. If it were true that a transaction could fail the block. Then we could be perpetually failing to generate a block in all rounds because the bad transaction would be gossiped around to all peers. This would be true already. Therefore it is much safer for our sanity to crash here because we should never be constructing a bad block ourselves.

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 5]
```diff
@@ -132,10 +132,9 @@ impl Sumeragi {
                 .control_message_receiver
                 .try_recv()
                 .map_err(|recv_error| {
-                    assert!(
-                        recv_error != mpsc::TryRecvError::Disconnected,
-                        "INTERNAL ERROR: Sumeragi control message pump disconnected"
-                    )
+                    if recv_error == mpsc::TryRecvError::Disconnected {
+                        error!("INTERNAL ERROR: Sumeragi control message pump disconnected");
+                    }
```
Comment (mversic) [inline crates/iroha_core/src/sumeragi/main_loop.rs:137]: I think panic was appropriate here, we need to shut down iroha

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 6]
```diff
@@ -171,10 +170,9 @@ impl Sumeragi {
                 .message_receiver
                 .try_recv()
                 .map_err(|recv_error| {
-                    assert!(
-                        recv_error != mpsc::TryRecvError::Disconnected,
-                        "INTERNAL ERROR: Sumeragi message pump disconnected"
-                    )
+                    if recv_error == mpsc::TryRecvError::Disconnected {
+                        error!("INTERNAL ERROR: Sumeragi message pump disconnected");
+                    }
```
Comment (mversic) [inline crates/iroha_core/src/sumeragi/main_loop.rs:175]: also here

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 7]
```diff
@@ -841,57 +838,74 @@ impl Sumeragi {
...
-            if created_in > pipeline_time / 2 {
-                warn!(
-                    role=%self.role(),
-                    peer_id=%self.peer_id,
-                    "Creating block takes too much time. \
-                    This might prevent consensus from operating. \
-                    Consider increasing `commit_time` or decreasing `max_transactions_in_block`"
-                );
+            let pre_signed_block = BlockPayload {
+                header: BlockBuilder::make_header(
```
Comment (mversic) [inline crates/iroha_core/src/sumeragi/main_loop.rs:856]: there's a `new` constructor. Please use it

File: crates/iroha_core/src/block.rs [hunk 8]
```diff
@@ -153,7 +153,8 @@ mod pending {
             Self(Pending { transactions })
         }
 
-        fn make_header(
+        /// Create a block header
+        pub fn make_header(
```
Comment (mversic) [inline crates/iroha_core/src/block.rs:157]:
```suggestion
       fn make_header(
```

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 9]
```diff
@@ -841,57 +838,74 @@ impl Sumeragi {
...
 
         if tx_cache_full || (deadline_reached && tx_cache_non_empty) {
             let transactions = self
                 .transaction_cache
                 .iter()
-                .map(|tx| tx.deref().clone())
+                .map(|tx| CommittedTransaction {
+                    value: tx.deref().clone().into(),
+                    error: None,
+                })
```
Comment (mversic) [inline crates/iroha_core/src/sumeragi/main_loop.rs:852]: this should happen inside `BlockBuilder::new`. Restore it and move this code there

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 10]
```diff
@@ -312,13 +313,11 @@ impl Sumeragi {
         // NOTE: By this time genesis block is executed and list of trusted peers is updated
         self.topology = Topology::new(state_block.world.trusted_peers_ids.clone());
 
-        let msg = BlockCreated::from(&genesis);
         let genesis = genesis
             .commit(&self.topology)
             .unpack(|e| self.send_event(e))
             .expect("Genesis invalid");
 
-        self.broadcast_packet(msg);
```
[the remark this answers is not in the document]
Comment (mversic) [reply crates/iroha_core/src/sumeragi/main_loop.rs:321]: yes, please restore this. It's better that all peers are notified immediately and block sync can take over later.
For genesis all transactions have `CommittedTransaction::error = None`

**Review by mversic — CHANGES_REQUESTED** [review]
I don't see that `CommittedTransaction::error` is being updated anywhere

File: crates/iroha_core/src/sumeragi/main_loop.rs [hunk 11]
```diff
@@ -841,57 +838,74 @@ impl Sumeragi {
             .expect("INTERNAL BUG: transactions in block exceed usize::MAX");
         let block_time = state.world.view().parameters.sumeragi.block_time();
         let tx_cache_full = self.transaction_cache.len() >= max_transactions.get();
-        let deadline_reached = self.round_start_time.elapsed() > block_time;
+        let deadline_reached =
+            self.topology.view_change_index() > 0 || self.round_start_time.elapsed() > block_time;
```
Comment (mversic) [inline crates/iroha_core/src/sumeragi/main_loop.rs:842]:
```suggestion
        let deadline_reached =
            self.topology.view_change_index() > 0 || self.round_start_time.elapsed() > block_time;
```

this is abuse of the word `deadline`. view change index is not related to deadline so keep it out

File: crates/iroha_core/src/sumeragi/main_loop.rs (hunk 3 above)
Comment (Erigara) [reply crates/iroha_core/src/sumeragi/main_loop.rs:894]: You can check `validate_keep_voting_block` and `validate_transactions` inside it.
If there is mismatch in transaction error (and we set it to `None` despite it's actual status) we got rejection.
It's not possible to fully validate transaction in the queue without executing it.

Comment (mversic) [reply crates/iroha_core/src/sumeragi/main_loop.rs:894]: We want to categorize transactions and reject a block if it's `CommittedTransaction::error` is incorrect from what the current peer's evaluation produced

Comment (mversic) [reply crates/iroha_core/src/sumeragi/main_loop.rs:894]: @erigara but how can there be a mismatch in the transaction error when the received block has it set to None for all transactions?

Comment (Erigara) [reply crates/iroha_core/src/sumeragi/main_loop.rs:894]: Look [here](https://github.com/hyperledger/iroha/blob/002803facb9d2ae9fb414f5428e6356510fe094c/crates/iroha_core/src/block.rs#L619), if transaction has `None` but after execution we discovered that it not actually `None` error is returned

Comment (mversic) [reply crates/iroha_core/src/sumeragi/main_loop.rs:894]: > Look [here](https://github.com/hyperledger/iroha/blob/002803facb9d2ae9fb414f5428e6356510fe094c/crates/iroha_core/src/block.rs#L619), if transaction has `None` but after execution we discovered that it not actually `None` error is returned   

Shouldn't we remove that check there?

Comment (SamHSmith) [reply crates/iroha_core/src/sumeragi/main_loop.rs:899]: I have removed that check and replaced it with a mutation of the error value in the block.

**Review by mversic — APPROVED** [review]

**Review by Erigara — APPROVED** [review]
